File: 23-dp2/3-dp-optimization/main.py
# DP Optimization - Python example (Space Optimized Fibonacci)

def fib(n):
    if n <= 1:
        return n
    prev, curr = 0, 1
    for _ in range(2, n+1):
        prev, curr = curr, prev + curr
    return curr

if __name__ == "__main__":
    print(fib(10))  # Output: 55

File: 23-dp2/3-dp-optimization/main.java
// DP Optimization - Java example (Space Optimized Fibonacci)

public class Main {
    public static int fib(int n) {
        if (n <= 1) return n;
        int prev = 0, curr = 1;
        for (int i = 2; i <= n; i++) {
            int next = prev + curr;
            prev = curr;
            curr = next;
        }
        return curr;
    }

    public static void main(String[] args) {
        System.out.println(fib(10));  // Output: 55
    }
}
